File: Autonomous_Mobile_Robots_Course/Chapter15/Lesson5/Chapter15_Lesson5.cpp
// Chapter15_Lesson5.cpp
// Lab: Compare Local Planners in Dense Obstacles (DWA vs simplified TEB-proxy)
// Build: g++ -O2 -std=c++17 Chapter15_Lesson5.cpp -o lab
// Run:   ./lab 30 0
#include <algorithm>
#include <array>
#include <cmath>
#include <iostream>
#include <random>
#include <vector>

struct Ob { double x,y,r; };
struct Res { bool success=false, collision=false; double time=0, L=0, cmin=1e9, w2=0; };

static inline double wrap(double a){ a = std::fmod(a + M_PI, 2*M_PI); if(a<0) a+=2*M_PI; return a - M_PI; }
static inline double hypot2(double x,double y){ return std::sqrt(x*x+y*y); }
static inline double corridor_y(double x){ return 0.9*std::sin(0.6*x); }

std::vector<std::array<double,2>> reference_path(int n=120){
  std::vector<std::array<double,2>> p; p.reserve(n);
  double x0=0.5,x1=11.5;
  for(int i=0;i<n;i++){
    double x = x0 + (x1-x0)*double(i)/(n-1);
    p.push_back({x, corridor_y(x)});
  }
  return p;
}

std::vector<Ob> sample_dense_world(int nObs, int seed){
  std::mt19937 rng(seed);
  std::uniform_real_distribution<double> Ux(0.5, 11.5), Uy(-2.5, 2.5), Ur(0.12, 0.32);
  std::vector<Ob> obs; obs.reserve(nObs);
  int tries=0;
  while((int)obs.size()<nObs && tries<20000){
    tries++;
    double x=Ux(rng), y=Uy(rng), r=Ur(rng);
    if(std::abs(y - corridor_y(x)) < 0.55 + r) continue;
    bool ok=true;
    for(auto &o: obs){
      double dx=x-o.x, dy=y-o.y;
      double rr=r+o.r+0.05;
      if(dx*dx+dy*dy < rr*rr){ ok=false; break; }
    }
    if(ok) obs.push_back({x,y,r});
  }
  return obs;
}

double clearance(double px,double py, const std::vector<Ob>& obs, double R){
  double d=1e18;
  for(auto &o: obs){
    d = std::min(d, hypot2(px-o.x, py-o.y) - o.r - R);
  }
  return d;
}

std::array<double,3> step_unicycle(const std::array<double,3>& x, double v,double w,double dt){
  return { x[0] + v*std::cos(x[2])*dt,
           x[1] + v*std::sin(x[2])*dt,
           wrap(x[2] + w*dt) };
}

std::pair<std::array<double,3>, double> rollout_min_clear(std::array<double,3> x, double v,double w,
                                                          const std::vector<Ob>& obs, double R, double dt, double T){
  int steps = std::max(1, (int)std::lround(T/dt));
  double mc=1e18;
  for(int i=0;i<steps;i++){
    x = step_unicycle(x,v,w,dt);
    mc = std::min(mc, clearance(x[0],x[1],obs,R));
    if(mc<=0) break;
  }
  return {x, mc};
}

struct Params{
  double R=0.25, vmax=0.9, wmax=1.6, a=1.2, alpha=2.5;
  double dt=0.1, T=2.0, goal_tol=0.25;
  int max_steps=800, n_obs=45;
};

std::pair<double,double> dwa_command(const std::array<double,3>& x, double v0,double w0,
                                     const std::vector<std::array<double,2>>& path,
                                     const std::vector<Ob>& obs, const Params& p){
  double vmin=std::max(0.0, v0 - p.a*p.dt), vmax=std::min(p.vmax, v0 + p.a*p.dt);
  double wmin=std::max(-p.wmax, w0 - p.alpha*p.dt), wmax=std::min(p.wmax, w0 + p.alpha*p.dt);

  double bestJ=-1e18, bestV=0, bestW=0;
  auto goal = path.back();

  for(int i=0;i<9;i++){
    double v = vmin + (vmax-vmin)*double(i)/8.0;
    for(int j=0;j<17;j++){
      double w = wmin + (wmax-wmin)*double(j)/16.0;
      auto rr = rollout_min_clear(x,v,w,obs,p.R,p.dt,p.T);
      double mc = rr.second;
      if(mc <= p.R + 0.05) continue;
      double gx = rr.first[0]-goal[0], gy = rr.first[1]-goal[1];
      double goalDist = hypot2(gx,gy);
      double J = -0.6*goalDist + 1.8*mc + 0.4*(v/(p.vmax+1e-9)) - 0.12*(w*w);
      if(J>bestJ){ bestJ=J; bestV=v; bestW=w; }
    }
  }
  return {bestV, bestW};
}

std::vector<std::array<double,2>> optimize_band(const std::array<double,3>& x,
                                                const std::vector<Ob>& obs, const Params& p,
                                                int N=12, double bandLen=2.6, int iters=20, double step=0.12){
  std::vector<std::array<double,2>> pts(N);
  for(int i=0;i<N;i++){
    double xx = x[0] + bandLen*double(i)/(N-1);
    pts[i] = {xx, corridor_y(xx)};
  }
  pts[0] = {x[0], x[1]};

  for(int it=0; it<iters; ++it){
    std::vector<std::array<double,2>> g(N, {0,0});

    // smoothness
    for(int i=1;i<N-1;i++){
      g[i][0] += 0.35*(2*pts[i][0] - pts[i-1][0] - pts[i+1][0]);
      g[i][1] += 0.35*(2*pts[i][1] - pts[i-1][1] - pts[i+1][1]);
    }

    // obstacle repulsion (nearest obstacle)
    for(int i=1;i<N;i++){
      double bestD=1e18; std::array<double,2> bestU{0,0};
      for(auto &o: obs){
        double vx=pts[i][0]-o.x, vy=pts[i][1]-o.y;
        double n = hypot2(vx,vy) + 1e-12;
        double d = n - o.r - p.R;
        if(d<bestD){ bestD=d; bestU={vx/n, vy/n}; }
      }
      if(bestD < 0.9){
        double phi = std::exp(-4.5*(bestD - 0.9));
        g[i][0] += 1.8*(-4.5*phi)*bestU[0];
        g[i][1] += 1.8*(-4.5*phi)*bestU[1];
      }
    }

    // time preference
    for(int i=1;i<N;i++){
      g[i][0] += 0.25*(pts[i][0]-pts[i-1][0]);
      g[i][1] += 0.25*(pts[i][1]-pts[i-1][1]);
    }

    for(int i=1;i<N;i++){
      pts[i][0] -= step*g[i][0];
      pts[i][1] -= step*g[i][1];
    }
    pts[0] = {x[0], x[1]};
  }
  return pts;
}

std::pair<double,double> teb_command(const std::array<double,3>& x, double v0,double w0,
                                     const std::vector<std::array<double,2>>& path,
                                     const std::vector<Ob>& obs, const Params& p){
  (void)path; // not needed in this proxy
  auto pts = optimize_band(x, obs, p);
  double dx=pts[1][0]-pts[0][0], dy=pts[1][1]-pts[0][1];
  double heading = std::atan2(dy,dx);
  double herr = wrap(heading - x[2]);

  double w = std::clamp(2.2*herr, -p.wmax, p.wmax);
  double v = std::clamp(0.8*(hypot2(dx,dy)/p.dt), 0.0, p.vmax);
  v *= 1.0/(1.0 + 1.2*std::abs(w));

  // accel limits
  v = std::clamp(v, std::max(0.0, v0 - p.a*p.dt), std::min(p.vmax, v0 + p.a*p.dt));
  w = std::clamp(w, w0 - p.alpha*p.dt, w0 + p.alpha*p.dt);
  return {v,w};
}

Res run_one(const std::string& planner, int seed, const Params& p){
  auto obs  = sample_dense_world(p.n_obs, seed);
  auto path = reference_path();
  auto goal = path.back();

  std::array<double,3> x{0.6,0.0,0.0};
  double v=0,w=0;
  Res r; r.time = p.max_steps*p.dt;

  for(int k=0;k<p.max_steps;k++){
    double c = clearance(x[0],x[1],obs,p.R);
    r.cmin = std::min(r.cmin, c);
    if(c<=0){ r.collision=true; r.time=k*p.dt; return r; }

    double gdist = hypot2(x[0]-goal[0], x[1]-goal[1]);
    if(gdist <= p.goal_tol){ r.success=true; r.time=k*p.dt; return r; }

    double vcmd, wcmd;
    if(planner=="teb"){ auto u = teb_command(x,v,w,path,obs,p); vcmd=u.first; wcmd=u.second; }
    else             { auto u = dwa_command(x,v,w,path,obs,p); vcmd=u.first; wcmd=u.second; }

    auto x2 = step_unicycle(x, vcmd, wcmd, p.dt);
    r.L  += hypot2(x2[0]-x[0], x2[1]-x[1]);
    r.w2 += (wcmd*wcmd)*p.dt;
    x=x2; v=vcmd; w=wcmd;
  }
  return r;
}

struct Summary{ int n=0; double succ=0,col=0, t=0,L=0,cmin=0,w2=0; };

Summary summarize(const std::vector<Res>& R){
  Summary s; s.n = (int)R.size();
  for(auto &r: R){
    s.succ += r.success?1:0;
    s.col  += r.collision?1:0;
    s.t    += r.time; s.L += r.L; s.cmin += r.cmin; s.w2 += r.w2;
  }
  s.t/=s.n; s.L/=s.n; s.cmin/=s.n; s.w2/=s.n;
  return s;
}

int main(int argc, char** argv){
  int trials = (argc>1)? std::atoi(argv[1]) : 30;
  int seed0  = (argc>2)? std::atoi(argv[2]) : 0;
  Params p;

  for(auto name: {std::string("dwa"), std::string("teb")}){
    std::vector<Res> R; R.reserve(trials);
    for(int i=0;i<trials;i++) R.push_back(run_one(name, seed0+i, p));
    auto s = summarize(R);
    std::cout << "\n" << (name=="dwa"?"DWA":"TEB") << " trials="<<s.n
              << " success_rate="<< (s.succ/s.n)
              << " collision_rate="<< (s.col/s.n)
              << " time_mean="<< s.t
              << " L_mean="<< s.L
              << " cmin_mean="<< s.cmin
              << " w2_mean="<< s.w2
              << "\n";
  }
  return 0;
}

File: Tutorials/Codes/Autonomous_Mobile_Robots_Course/Chapter15/Lesson5/Chapter15_Lesson5.java
// Chapter15_Lesson5.java
// Lab: Compare Local Planners in Dense Obstacles (DWA vs simplified TEB-proxy)
// Build: javac Chapter15_Lesson5.java
// Run:   java Chapter15_Lesson5 30 0
import java.util.*;

public class Chapter15_Lesson5 {

  static class Ob { double x,y,r; Ob(double x,double y,double r){this.x=x;this.y=y;this.r=r;} }
  static class Res { boolean success=false, collision=false; double time=0,L=0,cmin=1e9,w2=0; }
  static class Params{
    double R=0.25,vmax=0.9,wmax=1.6,a=1.2,alpha=2.5,dt=0.1,T=2.0,goalTol=0.25;
    int maxSteps=800,nObs=45;
  }

  static double wrap(double a){
    a = (a + Math.PI) % (2*Math.PI);
    if(a < 0) a += 2*Math.PI;
    return a - Math.PI;
  }
  static double hypot(double x,double y){ return Math.hypot(x,y); }
  static double corridorY(double x){ return 0.9*Math.sin(0.6*x); }

  static double[][] referencePath(int n){
    double x0=0.5,x1=11.5;
    double[][] p = new double[n][2];
    for(int i=0;i<n;i++){
      double x = x0 + (x1-x0)*((double)i/(n-1));
      p[i][0]=x; p[i][1]=corridorY(x);
    }
    return p;
  }

  static ArrayList<Ob> sampleDenseWorld(int nObs, int seed){
    Random rng = new Random(seed);
    ArrayList<Ob> obs = new ArrayList<>(nObs);
    int tries=0;
    while(obs.size()<nObs && tries<20000){
      tries++;
      double x = 0.5 + 11.0*rng.nextDouble();
      double y = -2.5 + 5.0*rng.nextDouble();
      double r = 0.12 + 0.20*rng.nextDouble();
      if(Math.abs(y - corridorY(x)) < 0.55 + r) continue;

      boolean ok=true;
      for(Ob o: obs){
        double dx=x-o.x, dy=y-o.y;
        double rr=r+o.r+0.05;
        if(dx*dx+dy*dy < rr*rr){ ok=false; break; }
      }
      if(ok) obs.add(new Ob(x,y,r));
    }
    return obs;
  }

  static double clearance(double px,double py, ArrayList<Ob> obs, double R){
    double d=1e18;
    for(Ob o: obs){
      d = Math.min(d, hypot(px-o.x, py-o.y) - o.r - R);
    }
    return d;
  }

  static double[] stepUnicycle(double[] x, double v,double w,double dt){
    return new double[]{ x[0] + v*Math.cos(x[2])*dt,
                         x[1] + v*Math.sin(x[2])*dt,
                         wrap(x[2] + w*dt) };
  }

  static class Rollout { double[] xf; double minClear; Rollout(double[] xf,double mc){this.xf=xf; this.minClear=mc;} }

  static Rollout rolloutMinClear(double[] x0, double v,double w, ArrayList<Ob> obs, Params p){
    int steps = Math.max(1, (int)Math.round(p.T/p.dt));
    double[] x = x0.clone();
    double mc = 1e18;
    for(int i=0;i<steps;i++){
      x = stepUnicycle(x,v,w,p.dt);
      mc = Math.min(mc, clearance(x[0],x[1],obs,p.R));
      if(mc<=0) break;
    }
    return new Rollout(x,mc);
  }

  static double[] dwaCommand(double[] x, double v0,double w0, double[][] path, ArrayList<Ob> obs, Params p){
    double vmin=Math.max(0.0, v0 - p.a*p.dt), vmax=Math.min(p.vmax, v0 + p.a*p.dt);
    double wmin=Math.max(-p.wmax, w0 - p.alpha*p.dt), wmax=Math.min(p.wmax, w0 + p.alpha*p.dt);
    double bestJ=-1e18, bestV=0, bestW=0;
    double[] goal = path[path.length-1];

    for(int i=0;i<9;i++){
      double v = vmin + (vmax-vmin)*(i/8.0);
      for(int j=0;j<17;j++){
        double w = wmin + (wmax-wmin)*(j/16.0);
        Rollout rr = rolloutMinClear(x,v,w,obs,p);
        double mc = rr.minClear;
        if(mc <= p.R + 0.05) continue;
        double gx=rr.xf[0]-goal[0], gy=rr.xf[1]-goal[1];
        double goalDist = hypot(gx,gy);
        double J = -0.6*goalDist + 1.8*mc + 0.4*(v/(p.vmax+1e-9)) - 0.12*(w*w);
        if(J>bestJ){ bestJ=J; bestV=v; bestW=w; }
      }
    }
    return new double[]{bestV,bestW};
  }

  static double[][] optimizeBand(double[] x, ArrayList<Ob> obs, Params p, int N, double bandLen, int iters, double step){
    double[][] pts = new double[N][2];
    for(int i=0;i<N;i++){
      double xx = x[0] + bandLen*(i/(double)(N-1));
      pts[i][0]=xx; pts[i][1]=corridorY(xx);
    }
    pts[0][0]=x[0]; pts[0][1]=x[1];

    for(int it=0; it<iters; it++){
      double[][] g = new double[N][2];

      for(int i=1;i<N-1;i++){
        g[i][0] += 0.35*(2*pts[i][0] - pts[i-1][0] - pts[i+1][0]);
        g[i][1] += 0.35*(2*pts[i][1] - pts[i-1][1] - pts[i+1][1]);
      }

      for(int i=1;i<N;i++){
        double bestD=1e18, ux=0, uy=0;
        for(Ob o: obs){
          double vx=pts[i][0]-o.x, vy=pts[i][1]-o.y;
          double n = hypot(vx,vy) + 1e-12;
          double d = n - o.r - p.R;
          if(d<bestD){ bestD=d; ux=vx/n; uy=vy/n; }
        }
        if(bestD < 0.9){
          double phi = Math.exp(-4.5*(bestD - 0.9));
          g[i][0] += 1.8*(-4.5*phi)*ux;
          g[i][1] += 1.8*(-4.5*phi)*uy;
        }
      }

      for(int i=1;i<N;i++){
        g[i][0] += 0.25*(pts[i][0]-pts[i-1][0]);
        g[i][1] += 0.25*(pts[i][1]-pts[i-1][1]);
      }

      for(int i=1;i<N;i++){
        pts[i][0] -= step*g[i][0];
        pts[i][1] -= step*g[i][1];
      }
      pts[0][0]=x[0]; pts[0][1]=x[1];
    }
    return pts;
  }

  static double[] tebCommand(double[] x, double v0,double w0, double[][] path, ArrayList<Ob> obs, Params p){
    double[][] pts = optimizeBand(x, obs, p, 12, 2.6, 20, 0.12);
    double dx=pts[1][0]-pts[0][0], dy=pts[1][1]-pts[0][1];
    double heading = Math.atan2(dy,dx);
    double herr = wrap(heading - x[2]);

    double w = Math.max(-p.wmax, Math.min(p.wmax, 2.2*herr));
    double v = Math.max(0.0, Math.min(p.vmax, 0.8*(hypot(dx,dy)/p.dt)));
    v *= 1.0/(1.0 + 1.2*Math.abs(w));

    v = Math.max(Math.max(0.0, v0 - p.a*p.dt), Math.min(Math.min(p.vmax, v0 + p.a*p.dt), v));
    w = Math.max(w0 - p.alpha*p.dt, Math.min(w0 + p.alpha*p.dt, w));
    return new double[]{v,w};
  }

  static Res runOne(String planner, int seed, Params p){
    ArrayList<Ob> obs = sampleDenseWorld(p.nObs, seed);
    double[][] path = referencePath(120);
    double[] goal = path[path.length-1];

    double[] x = new double[]{0.6,0.0,0.0};
    double v=0,w=0;
    Res r = new Res();
    r.time = p.maxSteps*p.dt;

    for(int k=0;k<p.maxSteps;k++){
      double c = clearance(x[0],x[1],obs,p.R);
      r.cmin = Math.min(r.cmin, c);
      if(c<=0){ r.collision=true; r.time=k*p.dt; return r; }
      if(hypot(x[0]-goal[0], x[1]-goal[1]) <= p.goalTol){ r.success=true; r.time=k*p.dt; return r; }

      double[] u = planner.equals("teb") ? tebCommand(x,v,w,path,obs,p) : dwaCommand(x,v,w,path,obs,p);
      double vcmd=u[0], wcmd=u[1];

      double[] x2 = stepUnicycle(x, vcmd, wcmd, p.dt);
      r.L += hypot(x2[0]-x[0], x2[1]-x[1]);
      r.w2 += (wcmd*wcmd)*p.dt;
      x=x2; v=vcmd; w=wcmd;
    }
    return r;
  }

  static Map<String,Double> summarize(ArrayList<Res> R){
    int n=R.size();
    double succ=0,col=0,t=0,L=0,cmin=0,w2=0;
    for(Res r: R){
      succ += r.success?1:0; col += r.collision?1:0;
      t += r.time; L += r.L; cmin += r.cmin; w2 += r.w2;
    }
    LinkedHashMap<String,Double> s = new LinkedHashMap<>();
    s.put("trials", (double)n);
    s.put("success_rate", succ/n);
    s.put("collision_rate", col/n);
    s.put("time_mean", t/n);
    s.put("L_mean", L/n);
    s.put("cmin_mean", cmin/n);
    s.put("w2_mean", w2/n);
    return s;
  }

  public static void main(String[] args){
    int trials = (args.length>0) ? Integer.parseInt(args[0]) : 30;
    int seed0  = (args.length>1) ? Integer.parseInt(args[1]) : 0;
    Params p = new Params();

    for(String name: new String[]{"dwa","teb"}){
      ArrayList<Res> R = new ArrayList<>(trials);
      for(int i=0;i<trials;i++) R.add(runOne(name, seed0+i, p));
      System.out.println("\n" + name.toUpperCase() + " " + summarize(R));
    }
  }
}
